presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.sketch.kll;
+
+import com.facebook.presto.common.array.ObjectBigArray;
+import com.facebook.presto.operator.aggregation.sketch.theta.ThetaSketchStateFactory;
+import com.facebook.presto.operator.aggregation.state.AbstractGroupedAccumulatorState;
+import com.facebook.presto.spi.function.AccumulatorState;
+import com.facebook.presto.spi.function.AccumulatorStateMetadata;
+import org.apache.datasketches.kll.KllItemsSketch;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+
+import static java.util.Objects.requireNonNull;
+
+@AccumulatorStateMetadata(stateFactoryClass = KllSketchStateFactory.class, stateSerializerClass = KllSketchStateSerializer.class)
+public interface KllSketchAggregationState
+        extends AccumulatorState
+{
+    @Nullable
+    <T> KllItemsSketch<T> getSketch();
+
+    <T> void setSketch(KllItemsSketch<T> sketch);
+
+    class Single
+            implements KllSketchAggregationState
+    {
+        private static final long INSTANCE_SIZE = ClassLayout.parseClass(Single.class).instanceSize();
+        @Nullable
+        private KllItemsSketch sketch;
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public <T> KllItemsSketch<T> getSketch()
+        {
+            return sketch;
+        }
+
+        @Override
+        public <T> void setSketch(KllItemsSketch<T> sketch)
+        {
+            this.sketch = sketch;
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            if (sketch != null) {
+                return sketch.getSerializedSizeBytes() + INSTANCE_SIZE;
```

Comment:
Is this the same thing as the size in memory?  If not, we need to use a more accurate method that captures the size in memory for accounting purposes.  If they're off, then such error can grow in magnitude for large group by or when the input is skewed.

Reply:
I always assumed it was the serialized size. Docs on the interface would be nice...

Anyways, there isn't going to be a perfect way to calculate the size, especially for variable-width types as we discussed offline. For fixed-width types, I think a heuristic might be okay. I replaced this with the following function:

```java
        static final long SKETCH_INSTANCE_SIZE = ClassLayout.parseClass(KllItemsSketch.class).instanceSize();
        @Override
        public long getEstimatedSize()
        {
            if (sketch != null) {
                // Estimating the memory is difficult without direct access to
                // the underlying implementation. We will attempt to give a
                // close estimate without underestimating to prevent crashes.
                // Overestimates might claim more resources, and prevent more
                // queries from running.
                // Since retained entries generally dominate the memory usage.
                return (SKETCH_INSTANCE_SIZE +
                        INSTANCE_SIZE +
                        (long) (sketch.getNumRetained() * retainedItemSize * 1.2));
            }

            return INSTANCE_SIZE;
        }
```

I ran some rough benchmarks using JOL's `GraphLayout#parseInstance` which attempts to accurately calculate retained memory of a single object. I compared it against the above java function to test the accuracy. Here are the results:

![image](https://github.com/prestodb/presto/assets/3464508/f255e856-ae18-4a67-b49e-eec262683eac)

![image](https://github.com/prestodb/presto/assets/3464508/97b79fb2-9610-43f2-a2e1-4ed513f2aec3)

At around 2048 items (2^11), the error percentage drops under 1 percent, which decreases up to 0.1% after the sketch has seen > 100k items.
I might be able to modify the heuristic to account for sketches which have seen a smaller number of item. from the stream. But I think the error is within a reasonable margin? What are your thoughts on the reasonable margin of error?

Reply:
FWIW I also experimented against the T-Digest using a compression factor giving comparable sizes to the above chart:
![image](https://github.com/prestodb/presto/assets/3464508/c6734bde-3f6c-467d-aa25-014d172ed5a4)

The maximum error here is roughly 20%, while the minimum is ~0.27%

Reply:
@ZacBlanco although this seems like a moderate risk, I am wary of doing something Presto-side since upgrades might be unstable and cause the heuristic to become out much more inaccurate.

Can you estimate how does the serialized size compare to JOL's estimation?  If they're close, then perhaps we could use that (plus a fudge factor), as it seems less likely that the data structure would change without a corresponding change to the serialized layout.  Additionally, you could update the grouped version to keep track one-by-one using the same heuristic.

Reply:
Sorry for the delay on this one. I spent some time analyzing to come up with a decent heuristic that
we can use for the meantime based on the serialized size of the sketch.

While it's possible to make an addition to the sketch APIs to report utilized memory it's difficult
as they make heavy use of abstract classes which makes the calculation more complex than
most of the `ClassLayout#parseClass` instances in the Presto codebase. I started a small
implementation that might work for our case but proposing and implementing that API will probably
take some time to get accepted by the community.

While we wait for that change to land in the datasketches upstream, I ran a regression analysis to
determine the best adjustment factor to use. For each datatype, you can model the true in-memory
size, $t$ as $t(x) = f_1 x + f_2$ where $x$ is the size of the serialized sketch. I think this
actually models the true size quite well because when serialized, the size will be dominated by the
number of retained items. But those retained items take up less space serialized than when in the
JVM object always by a constant factor (e.g. an instance of `Double` takes up 24 bytes, but when
serialized just 8 bytes). Any other overhead in the sketch is constant (e.g. min/max, seen count, etc)
which is why we need the constant factor adjustment. This method should work for both fixed-width
and variable-sized datatypes since the serialized size actually contains the values which contribute
to memory consumption - unlike the approach with retained items I tried above.

Using this model I ran the regression to find the best parameters of $f_1$ and $f_2$ for each
datatype. Some charts are below. For strings and numeric types the regression function seems to
pretty closely track the true usage. For booleans there is a discrepancy where the regression over-
estimates by a bit.

![image](https://github.com/prestodb/presto/assets/3464508/d117b790-01ce-4e75-af9b-1a82af48a6ed)

These were the parameters I found:

| Type | $f_1$ | $f_2$ |
|-----|-----|----|
| Double |  929.0746716729514 | 3.3328879969131457 |
| Long | 1008.5483794752541 | 4.234749475485517 |
| String | 345.60095530779705 | 14.12193288555095
| Boolean | 929.920850240635 | 3.3325974295999283 |

I would probably encode them into a static hashmap that gets referenced when querying the estimated size.

The ideal solution is for sure to get something upstream, but if we really care about getting accurate sizes for grouped queries, then this will be a pretty good stopgap solution with minimal error until then. The accuracy issues with booleans I am not very concerned about because there isn't really a great use case for booleans and KLL sketches anyways.

Reply:
Thank you for working on this detailed analysis.  I agree that your heuristic seems to be a good stopgap for now.

Reply:
The code has been updated with this approach